front/src/applications/operationalStudies/components/ImportTrainSchedule/ImportTrainScheduleConfig.tsx

```diff
@@ -30,25 +50,143 @@ export default function ImportTrainScheduleConfig(props: ImportTrainScheduleConf
+        !trainSchedule.rollingStock ||
+        !trainSchedule.departureTime ||
+        !trainSchedule.arrivalTime ||
+        !trainSchedule.departure ||
+        !trainSchedule.steps ||
+        !Array.isArray(trainSchedule.steps)
+      ) {
+        return true;
+      }
+      const hasInvalidateSteps = some(
+        trainSchedule.steps,
+        (step) =>
+          !('arrivalTime' in step) ||
+          !('departureTime' in step) ||
+          !('uic' in step) ||
+          // !('yard' in step) ||
+          !('name' in step) ||
+          !('trigram' in step) ||
+          !('latitude' in step) ||
+          !('longitude' in step)
+      );
+      return hasInvalidateSteps;
+    });
+    if (isInvalidTrainSchedules) {
+      dispatch(
+        setFailure({
+          name: t('errorMessages.error'),
+          message: 'Impossible de convertir les données en TrainSchedule',
+        })
+      );
+      console.error(
+        'Invalid response format: can not convert response into TrainSchedules. Expected response : { { trainNumber: string; rollingStock: string; departureTime: string; arrivalTime: string; departure: string; steps: ({uic: number; yard: string; name: string; trigram: string; latitude: number; longitude: number; arrivalTime: string; departureTime: string; })[]; transilienName?: string; }}'
+      );
+      return null;
+    }
+    return importedTrainSchedules as ImportedTrainSchedule[];
+  }
+
+  async function updateTrainSchedules(importedTrainSchedules: ImportedTrainSchedule[]) {
+    const opIds = importedTrainSchedules.flatMap((trainSchedule) =>
+      trainSchedule.steps.map((step) => step.uic)
+    );
+    const trackSectionsByOp = await importTrackSections(opIds);
+
+    // For each train schedule, we add the duration and tracks of each step
+    const trainsSchedules = importedTrainSchedules.map((trainSchedule) => {
+      const stepsWithDuration = trainSchedule.steps.map((step) => {
+        // calcul duration in seconds between step arrival and departure
+        // in case of arrival and departure are the same, we set duration to 0
+        // for the step arrivalTime is before departureTime because the train first goes to the station and then leaves it
+        const duration = Math.round(
+          (new Date(step.departureTime).getTime() - new Date(step.arrivalTime).getTime()) / 1000
+        );
+        return {
+          ...step,
+          duration,
+          tracks: trackSectionsByOp[Number(step.uic)] || [],
+        } as Step;
+      });
+      return {
+        ...trainSchedule,
+        steps: stepsWithDuration,
+      } as TrainSchedule;
+    });
+
+    setTrainsList(trainsSchedules);
+  }
+
+  async function getTrainsFromOpenData(config: TrainScheduleImportConfig) {
+    setTrainsList([]);
+    setIsLoading(true);
+    try {
+      const params = {
+        q: 'trains',
+        config,
+      };
+      const result = await get(`${GRAOU_URL}/api/trainschedules.php`, { params });
```

Comment:
this should not appear in a component

Reply:
I can create a function in a separate file :+1: